File: web/console/config/resource/k8sConfig/gateway.ts
/*
 * Tencent is pleased to support the open source community by making TKEStack
 * available.
 *
 * Copyright (C) 2012-2021 Tencent. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of the
 * License at
 *
 * https://opensource.org/licenses/Apache-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations under the License.
 */
import { ResourceInfo, DetailField, DisplayField, DetailInfo } from '../../../src/modules/common/models';
import { commonActionField, defaulNotExistedValue, apiVersion, dataFormatConfig } from '../common';
import { t, Trans } from '@tencent/tea-app/lib/i18n';

const displayField: DisplayField = {
  name: {
    dataField: ['metadata.name'],
    dataFormat: dataFormatConfig['text'],
    width: '20%',
    headTitle: t('名称'),
    noExsitedValue: defaulNotExistedValue,
    isLink: true, // 用于判断该值是否为链接
    isClip: true
  },
  servers: {
    dataField: ['servers'],
    dataFormat: dataFormatConfig['array'],
    width: '15%',
    headTitle: t('端口配置'),
    noExsitedValue: defaulNotExistedValue
  },
  selector: {
    dataField: ['selecor'],
    dataFormat: dataFormatConfig['text'],
    width: '25%',
    headTitle: t('类型'),
    noExsitedValue: defaulNotExistedValue
  },
  ip: {
    dataField: ['ip'],
    dataFormat: dataFormatConfig['text'],
    width: '25%',
    headTitle: t('IP'),
    noExsitedValue: '-'
  },
  creationTimestamp: {
    dataField: ['metadata.creationTimestamp'],
    dataFormat: dataFormatConfig['time'],
    width: '25%',
    headTitle: t('创建时间'),
    noExsitedValue: defaulNotExistedValue
  },
  operator: {
    dataField: [''],
    dataFormat: dataFormatConfig['operator'],
    width: '10%',
    headTitle: t('操作'),
    operatorList: [
      {
        name: t('删除'),
        actionType: 'delete',
        isInMoreOp: false
      }
    ]
  }
};

/** resrouce action当中的配置 */
const actionField = Object.assign({}, commonActionField);

/** 自定义tabList */
const tabList = [
  {
    id: 'nsInfo',
    label: t('详情')
  },
  {
    id: 'yaml',
    label: 'YAML'
  }
];

/** 自定义配置详情的展示 */
const detailBasicInfo: DetailInfo = {};

/** 详情页面的相关配置 */
const detailField: DetailField = {
  tabList,
  detailInfo: Object.assign({}, detailBasicInfo)
};

/** namespace的配置 */
export const gateway = (k8sVersion: string) => {
  // apiVersion的配置
  const apiKind = apiVersion[k8sVersion].gateway;
  let config: ResourceInfo = {
    headTitle: apiKind.headTitle,
    basicEntry: apiKind.basicEntry,
    group: apiKind.group,
    version: apiKind.version,
    namespaces: 'namespaces',
    requestType: {
      list: 'gateways'
    },
    displayField,
    actionField,
    detailField
  };
  return config;
};

File: pkg/gateway/gateway.go
/*
 * Tencent is pleased to support the open source community by making TKEStack
 * available.
 *
 * Copyright (C) 2012-2019 Tencent. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of the
 * License at
 *
 * https://opensource.org/licenses/Apache-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations under the License.
 */

package gateway

import (
	"bytes"
	"io/ioutil"
	"net/http"
	"regexp"

	"golang.org/x/oauth2"
	genericapiserver "k8s.io/apiserver/pkg/server"
	"tkestack.io/tke/pkg/apiserver/authentication/authenticator/oidc"
	"tkestack.io/tke/pkg/gateway/api"
	gatewayconfig "tkestack.io/tke/pkg/gateway/apis/config"
	"tkestack.io/tke/pkg/gateway/assets"
	"tkestack.io/tke/pkg/gateway/proxy"
	"tkestack.io/tke/pkg/gateway/websocket"
	"tkestack.io/tke/pkg/gateway/webtty"

	"html/template"
)

const (
	DefaultTitle   = "TKEStack"
	DefaultLogoDir = "default"
)

// ExtraConfig contains the additional configuration of apiserver.
type ExtraConfig struct {
	ServerName        string
	OAuthConfig       *oauth2.Config
	OIDCHttpClient    *http.Client
	OIDCAuthenticator *oidc.Authenticator
	GatewayConfig     *gatewayconfig.GatewayConfiguration
	HeaderRequest     bool
}

// Config contains the core configuration instance of server and additional
// configuration.
type Config struct {
	GenericConfig *genericapiserver.RecommendedConfig
	ExtraConfig   ExtraConfig
}

type completedConfig struct {
	GenericConfig genericapiserver.CompletedConfig
	ExtraConfig   *ExtraConfig
}

// CompletedConfig embed a private pointer of Config.
type CompletedConfig struct {
	// Embed a private pointer that cannot be instantiated outside of this package.
	*completedConfig
}

// Gateway contains state for TKE gateway server.
type Gateway struct {
	GenericAPIServer *genericapiserver.GenericAPIServer
}

// Complete fills in any fields not set that are required to have valid data.
// It's mutating the receiver.
func (cfg *Config) Complete() CompletedConfig {
	c := completedConfig{
		cfg.GenericConfig.Complete(),
		&cfg.ExtraConfig,
	}

	return CompletedConfig{&c}
}

// New returns a new instance of APIServer from the given config.
func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget) (*Gateway, error) {
	s, err := c.GenericConfig.New(c.ExtraConfig.ServerName, delegationTarget)
	if err != nil {
		return nil, err
	}

	registerCallbackRoute(s.Handler.NonGoRestfulMux, c.ExtraConfig.OAuthConfig, c.ExtraConfig.OIDCHttpClient, c.ExtraConfig.GatewayConfig.DisableOIDCProxy)

	if !c.ExtraConfig.GatewayConfig.DisableOIDCProxy {
		if err := registerAuthRoute(s.Handler.NonGoRestfulMux, c.ExtraConfig.OIDCHttpClient, c.ExtraConfig.OIDCAuthenticator); err != nil {
			return nil, err
		}
	}
	consoleConfig := new(gatewayconfig.ConsoleConfig)

	if c.ExtraConfig.GatewayConfig.ConsoleConfig != nil {
		consoleConfig = c.ExtraConfig.GatewayConfig.ConsoleConfig
	} else {
		consoleConfig.Title = DefaultTitle
		consoleConfig.LogoDir = DefaultLogoDir

	}

	files, err := ioutil.ReadDir(assets.RootDir)
	if err != nil {
		return nil, err
	}

	sourceRe := regexp.MustCompile(`\.tmpl\.html$`)
	targetRe := regexp.MustCompile(`\.tmpl`)

	for _, file := range files {
		if !sourceRe.MatchString(file.Name()) {
			continue
		}
		var buf bytes.Buffer
		t, err := template.New(file.Name()).Delims("{%", "%}").ParseFiles(assets.RootDir + file.Name())
		if err != nil {
			return nil, err
		}
		if err = t.Execute(&buf, consoleConfig); err != nil {
			return nil, err
		}
		// // remove .tmpl in file name
		targetFileName := targetRe.ReplaceAllString(file.Name(), "")
		if err = ioutil.WriteFile(assets.RootDir+targetFileName, buf.Bytes(), 0644); err != nil {
			return nil, err
		}
	}

	if err := proxy.RegisterRoute(s.Handler.NonGoRestfulMux, c.ExtraConfig.GatewayConfig, c.ExtraConfig.OIDCAuthenticator); err != nil {
		return nil, err
	}

	if err := api.RegisterRoute(s.Handler.GoRestfulContainer, c.ExtraConfig.GatewayConfig, c.ExtraConfig.OAuthConfig, c.ExtraConfig.OIDCHttpClient, c.ExtraConfig.OIDCAuthenticator, c.ExtraConfig.HeaderRequest); err != nil {
		return nil, err
	}

	if err := webtty.RegisterRoute(s.Handler.NonGoRestfulMux, c.ExtraConfig.GatewayConfig); err != nil {
		return nil, err
	}
	if err := websocket.RegisterRoute(s.Handler.NonGoRestfulMux, c.ExtraConfig.GatewayConfig); err != nil {
		return nil, err
	}

	if !c.ExtraConfig.HeaderRequest {
		assets.RegisterRoute(s.Handler.NonGoRestfulMux, c.ExtraConfig.OAuthConfig, c.ExtraConfig.GatewayConfig.DisableOIDCProxy)
	} else {
		assets.RegisterRoute(s.Handler.NonGoRestfulMux, nil, c.ExtraConfig.GatewayConfig.DisableOIDCProxy)
	}

	m := &Gateway{
		GenericAPIServer: s,
	}

	return m, nil
}
